packages/shared/src/pluginStore.ts

```diff
@@ -17,47 +82,9 @@ export function createPluginStore(): PluginStore {
     const { before } = options || {};
     for (const plugin of newPlugins) {
       if (!plugin) continue;
-      if (typeof plugin !== 'object') {
-        throw new Error(
-          `expect plugin instance is object, but got ${typeof plugin}.`,
-        );
-      }
-
-      const { name, setup, apply } = plugin as any;
-      if (typeof name !== 'string') {
-        logger.warn(
-          `Rsbuild plugin must have a "name" field for identification.`,
-        );
-      }
-      if (!isFunction(setup)) {
-        if (isFunction(apply)) {
-          const _name = plugin.constructor?.name ?? 'SomeWebpackPlugin';
-
-          const message =
-            `${color.yellow(_name)} looks like a webpack or Rspack plugin, ` +
-            `please use ${color.yellow('`tools.rspack`')} to register it:\n`;
-          const example = [
-            '  // rsbuild.config.ts',
-            '  export default {',
-            '    tools: {',
-            '      rspack: {',
-            `        plugins: [new ${name}()]`,
-            '      }',
-            '    }',
-            '  };',
-          ].join('\n');
-
-          throw new Error(message + color.green(example));
-        }
-
-        throw new Error(
-          `Expect Rsbuild plugin.setup to be a function, but got ${typeof plugin}.`,
-        );
-      }
+      assertPlugin(plugin);
       if (plugins.some((item) => item.name === plugin.name)) {
-        logger.warn(
-          `Rsbuild plugin "${plugin.name}" registered multiple times.`,
-        );
+        logger.warn(new PluginNameDuplicatedError(plugin));
```

Comment:
why use logger.warn to print an error?

Reply:
I just want to extract the error initialization to make the logic of `assertPlugin` clearer. Besides being used for pluginStore, I think it can also work in other cases where type guards are needed to restrict plugin types. 
Whether or not to display a warning when showing is worth discussing. Perhaps we can consider hiding the error stack by default in logger.warn, but displaying it when `LOG_LEVEL=verbose`.
Of course, this is not something that should be done in this PR. I would prefer it to be changed here as `logger.warn('STRING')`.